Require passing an `AttrWrapper` to `collect_tokens_trailing_token`

Aaron1011 opened the pull request:
This is a pure refactoring split out from #80689.
It represents the most invasive part of that PR, requiring changes in
every caller of `parse_outer_attributes`
    
In order to eagerly expand `#[cfg]` attributes while preserving the
original `TokenStream`, we need to know the range of tokens that
corresponds to every attribute target. This is accomplished by making
`parse_outer_attributes` return an opaque `AttrWrapper` struct. An
`AttrWrapper` must be converted to a plain `AttrVec` by passing it to
`collect_tokens_trailing_token`. This makes it difficult to accidentally
construct an AST node with attributes without calling `collect_tokens_trailing_token`,
since AST nodes store an `AttrVec`, not an `AttrWrapper`.
    
As a result, we now call `collect_tokens_trailing_token` for attribute
targets which only support inert attributes, such as generic arguments
and struct fields. Currently, the constructed `LazyTokenStream` is
simply discarded. Future PRs will record the token range corresponding
to the attribute target, allowing those tokens to be removed from an
enclosing `collect_tokens_trailing_token` call if necessary.

rust-highfive commented:
r? @lcnr

(rust-highfive has picked a reviewer for you, use r? to override)

Aaron1011 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

bors commented:
:hourglass: Trying commit 31748687020643dfc694f3dc9c70a6021c6b0b54 with merge 7db61ae90bf5c11a0be4801fa0c76f791029c5b4...
<!-- homu: {"type":"TryBuildStarted","head_sha":"31748687020643dfc694f3dc9c70a6021c6b0b54","merge_sha":"7db61ae90bf5c11a0be4801fa0c76f791029c5b4"} -->

lcnr commented:
r? @petrochenkov

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1751735625)
Build commit: 7db61ae90bf5c11a0be4801fa0c76f791029c5b4 (`7db61ae90bf5c11a0be4801fa0c76f791029c5b4`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1751735625"},"merge_sha":"7db61ae90bf5c11a0be4801fa0c76f791029c5b4"} -->

rust-timer commented:
Queued 7db61ae90bf5c11a0be4801fa0c76f791029c5b4 with parent b814b639836aa76b5c6deaa585367150bb2debf4, future [comparison URL](https://perf.rust-lang.org/compare.html?start=b814b639836aa76b5c6deaa585367150bb2debf4&end=7db61ae90bf5c11a0be4801fa0c76f791029c5b4).

@rustbot label: +S-waiting-on-perf

rust-timer commented:
Finished benchmarking try commit (7db61ae90bf5c11a0be4801fa0c76f791029c5b4): [comparison url](https://perf.rust-lang.org/compare.html?start=b814b639836aa76b5c6deaa585367150bb2debf4&end=7db61ae90bf5c11a0be4801fa0c76f791029c5b4).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf

petrochenkov commented:
I'm unlikely to get to reviewing this until the next weekend (Feb 7).

petrochenkov commented on compiler/rustc_parse/src/parser/stmt.rs:
```diff
@@ -38,30 +40,47 @@ impl<'a> Parser<'a> {
+                            mem::swap(stmt_attrs, &mut attrs);
+                            stmt_attrs.extend(attrs);
+                        });
+                        // Make sure we capture the token::Interpolated
```
Is this necessary?
The interpolated token is going to always be replaced by its underlying tokens.

petrochenkov replied:
Same applies to `parse_item_common`.

Aaron1011 replied:
Currently, I don't think this is strictly necessary, since we'll end up with the same attributes and tokens even if we don't call `collect_tokens_trailing_token`.

However, this will become necessary when we add `PreexpTokenStream` - calling `collect_tokens_trailing_token` will ensure that we properly insert any new attributes into the `PreexpTokenStream`. That is, we need `#[cfg(FALSE)] $item` to get removed from the `PreexpTokenStream`.

Alternatively, we could remove this logic, and directly manipulate the `PreexpTokenStream` (once it gets added). However, I think it's cleaner to have all of the token-related attribute handling go through `collect_tokens_trailing_token`.

petrochenkov commented on compiler/rustc_parse/src/parser/pat.rs:
```diff
@@ -938,16 +938,24 @@ impl<'a> Parser<'a> {
+                    }?;
+                    ate_comma = this.eat(&token::Comma);
+                    // We just ate a comma, so there's no need to use
+                    // `TrailingToken::Comma`
```
I think we should migrate to this scheme everywhere and eliminate `TrailingToken` entirely (not in this PR).

Aaron1011 replied:
The main obstacle to doing this will be `parse_seq_to_before_token` - it relies on being able to eat the separating tokens, and performs some non-trivial recovery logic. However, I have no objection to removing `TrailingToken`.

petrochenkov commented on compiler/rustc_parse/src/parser/nonterminal.rs:
```diff
@@ -105,7 +105,9 @@ impl<'a> Parser<'a> {
             },
             NonterminalKind::Block => {
-                token::NtBlock(self.collect_tokens(|this| this.parse_block())?)
+                // While an block *expression* may have attributes (e.g. `#[my_attr] { ... }`),
```
```suggestion
                // While a block *expression* may have attributes (e.g. `#[my_attr] { ... }`),
```

Aaron1011 commented on compiler/rustc_parse/src/parser/expr.rs:
```diff
@@ -457,44 +462,68 @@ impl<'a> Parser<'a> {
+                make_it!(this, attrs, |this, _| this.parse_box_expr(lo))
+            }
+            token::Ident(..) if this.is_mistaken_not_ident_negation() => {
+                // FIXME - what is our polciy for handling tokens during recovery?
```
I was thinking of https://github.com/rust-lang/rust/issues/76360#issuecomment-687639790, where recovery determines whether or not we invoke a proc-macro attribute at all. However, this case seems fine.

Aaron1011 commented on compiler/rustc_parse/src/parser/expr.rs:
```diff
@@ -457,44 +462,68 @@ impl<'a> Parser<'a> {
+            ($this:ident, $attrs:expr, |this, _| $body:expr) => {
+                $this.collect_tokens_for_expr($attrs, |$this, attrs| {
+                    let (hi, ex) = $body?;
+                    Ok($this.mk_expr(lo.to(hi), ex, attrs.into()))
```
We need to return an `Expr` from the closure, so that `collect_tokens_for_expr` can store the tokens via the `HasTokens` impl.

Aaron1011 commented:
@petrochenkov: I've addressed your comments. I've also moved `AttrWrapper` and some related functions to a new module, so that we actually enforce the privacy of `AttrWrapper.attrs`

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1894400817) [(plain)](https://github.com/rust-lang/rust/commit/474d1a6103e72b60053e27e97d5aa2ad8382f482/checks/1894400817/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/compiler/rustc_parse/src/parser/attr_wrapper.rs at line 1:
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_parse/src/parser/attr_wrapper.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
-use super::{Parser, TokenCursor, TrailingToken, ForceCollect};
 use super::attr;
-use rustc_ast::{self as ast};
+use super::{ForceCollect, Parser, TokenCursor, TrailingToken};
 use rustc_ast::token::{self, Token, TokenKind};
-use rustc_ast::tokenstream::{DelimSpan, Spacing, LazyTokenStream};
 use rustc_ast::tokenstream::{CreateTokenStream, TokenStream, TokenTree, TreeAndSpacing};
+use rustc_ast::tokenstream::{DelimSpan, LazyTokenStream, Spacing};
 use rustc_ast::HasTokens;
+use rustc_ast::{self as ast};
 use rustc_errors::PResult;
 use rustc_span::{Span, DUMMY_SP};
 
Diff in /checkout/compiler/rustc_parse/src/parser/attr_wrapper.rs at line 138:
         ret.finalize_tokens(LazyTokenStream::new(lazy_impl));
         Ok(ret)
-
 }
 
 
 /// Converts a flattened iterator of tokens (including open and close delimiter tokens)
Build completed unsuccessfully in 0:00:21
```

</details>

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit 3321d701617d5ba3ef348d9273f5fcca126d8f04 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 3321d701617d5ba3ef348d9273f5fcca126d8f04 -->
<!-- homu: {"type":"Approved","sha":"3321d701617d5ba3ef348d9273f5fcca126d8f04","approver":"petrochenkov"} -->

bors commented:
:hourglass: Testing commit 3321d701617d5ba3ef348d9273f5fcca126d8f04 with merge 5fa22fe6f821ac3801d05f624b123dda25fde32c...
<!-- homu: {"type":"BuildStarted","head_sha":"3321d701617d5ba3ef348d9273f5fcca126d8f04","merge_sha":"5fa22fe6f821ac3801d05f624b123dda25fde32c"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1898590451)
Approved by: petrochenkov
Pushing 5fa22fe6f821ac3801d05f624b123dda25fde32c to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1898590451"},"merge_sha":"5fa22fe6f821ac3801d05f624b123dda25fde32c"} -->
